ow_regolith/src/ModelPool.cpp

```diff
@@ -0,0 +1,174 @@
+  return
+    m_gz_spawn_model.connect<SpawnModel>(
+      m_node_handle, SRV_SPAWN_MODEL, SERVICE_CONNECT_TIMEOUT, true) &&
+    m_gz_delete_model.connect<DeleteModel>(
+      m_node_handle, SRV_DELETE_MODEL, SERVICE_CONNECT_TIMEOUT, true) &&
+    m_gz_apply_wrench.connect<ApplyBodyWrench>(
+      m_node_handle, SRV_APPLY_BODY_WRENCH, SERVICE_CONNECT_TIMEOUT, true) &&
+    m_gz_clear_wrench.connect<BodyRequest>(
+      m_node_handle, SRV_CLEAR_BODY_WRENCH, SERVICE_CONNECT_TIMEOUT, true);
+}
+
+bool ModelPool::setModel(const string &model_uri, const string &model_tag)
+{
+  // use local variables for data and only set members if all calls succeed
+  string model_sdf;
+  string model_body_name;
+  float model_mass;
+
+  // load SDF model
+  if (!getSdfFromUri(model_uri, model_sdf)) {
+    ROS_ERROR("Failed to load model SDF");
+    return false;
+  }
+
+  // parse as SDF object to query properties about the model
+  auto sdf = parseSdf(model_sdf);
+  if (!getModelLinkName(sdf, model_body_name)) {
+    ROS_ERROR("Failed to acquire model name");
+    return false;
+  }
+  if (!getModelLinkMass(sdf, model_mass)) {
+    ROS_ERROR("Failed to acquire model mass");
+    return false;
+  }
+
+  m_model_uri = model_uri;
+  m_model_sdf = model_sdf;
+  m_model_body_name = model_body_name;
+  m_model_tag = model_tag;
+  m_model_mass = model_mass;
+
+  return true;
+}
+
+string ModelPool::spawn(const Point &position, const string &reference_frame)
+{
+  ROS_INFO("Spawning regolith");
+
+  static auto spawn_count = 0;
+  stringstream model_name, link_name;
+  model_name << m_model_tag << "_" << spawn_count++;
+  link_name << model_name.str() << "::" << m_model_body_name;
+
+  SpawnModel spawn_msg;
+
+  spawn_msg.request.model_name      = model_name.str();
+  spawn_msg.request.model_xml       = m_model_sdf;
+  spawn_msg.request.reference_frame = reference_frame;
+
+  pointTFToMsg(position, spawn_msg.request.initial_pose.position);
+
+  if (!m_gz_spawn_model.call(spawn_msg))
+    return "";
+
+  m_active_models.emplace(
+    link_name.str(),
+    Model{model_name.str(), m_model_body_name}
+  );
+
+  return link_name.str();
+}
+
+vector<string> ModelPool::remove(const vector<string> &link_names)
+{
+  vector<string> not_removed;
+  DeleteModel msg;
+  if (link_names.empty()) {
+    // delete all regolith models
+    auto it = begin(m_active_models);
+    while (it != end(m_active_models)) {
+      msg.request.model_name = it->second.model_name;
+      if (removeModel(msg)) {
+        it = m_active_models.erase(it);
+      } else {
+        // do not remove from active list if removal failed
+        not_removed.push_back((it++)->first);
+      }
+    }
+  } else {
+    // delete one or multiple models as provided in link_names
+    for (auto const &n : link_names) {
+      try {
+        msg.request.model_name = m_active_models.at(n).model_name;
+      } catch(out_of_range &_err) {
+        not_removed.push_back(n);
+        continue;
+      }
```

Comment:
This seems logically incorrect, if **at** method failed this means that no such model exist, and thus it is not correct to consider it as **not_removed**

Reply:
Or maybe it is not clear to me how **not_removed** is used.

Reply:
**not_removed** is always a subset of **link_names** that were not removed for any reason. That reason can be either that the removeModel method failed, or that the link name does not exist inside this model pool. It does not differentiate between these reasons. The list enables whatever calls the service to perform more detailed error checking.